app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStorage.kt

```diff
@@ -0,0 +1,181 @@
+/*
+ * SPDX-FileCopyrightText: 2024 The Calyx Institute
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.stevesoltys.seedvault.plugins.webdav
+
+import android.util.Log
+import at.bitfire.dav4jvm.BasicDigestAuthHandler
+import at.bitfire.dav4jvm.DavCollection
+import at.bitfire.dav4jvm.Response
+import at.bitfire.dav4jvm.property.ResourceType
+import kotlinx.coroutines.DelicateCoroutinesApi
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import okhttp3.ConnectionSpec
+import okhttp3.HttpUrl
+import okhttp3.MediaType.Companion.toMediaType
+import okhttp3.OkHttpClient
+import okhttp3.RequestBody
+import okhttp3.internal.closeQuietly
+import okio.BufferedSink
+import java.io.IOException
+import java.io.InputStream
+import java.io.OutputStream
+import java.io.PipedInputStream
+import java.io.PipedOutputStream
+import java.util.concurrent.TimeUnit
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+const val DEBUG_LOG = true
+const val DIRECTORY_ROOT = ".SeedVaultAndroidBackup"
+
+@OptIn(DelicateCoroutinesApi::class)
+internal abstract class WebDavStorage(
+    webDavConfig: WebDavConfig,
+    root: String = DIRECTORY_ROOT,
+) {
+
+    companion object {
+        val TAG: String = WebDavStorage::class.java.simpleName
+    }
+
+    private val authHandler = BasicDigestAuthHandler(
+        domain = null, // Optional, to only authenticate against hosts with this domain.
+        username = webDavConfig.username,
+        password = webDavConfig.password,
+    )
+    protected val okHttpClient = OkHttpClient.Builder()
+        .followRedirects(false)
+        .authenticator(authHandler)
+        .addNetworkInterceptor(authHandler)
+        .connectTimeout(15, TimeUnit.SECONDS)
+        .writeTimeout(30, TimeUnit.SECONDS)
+        .readTimeout(120, TimeUnit.SECONDS)
+        .pingInterval(45, TimeUnit.SECONDS)
+        .connectionSpecs(listOf(ConnectionSpec.MODERN_TLS))
+        .retryOnConnectionFailure(true)
```

Comment:
Do we need to specify a retry limit? Is the default sufficient?

Reply:
so far I've found the default to work nicely. But we can tweak this as the need arises. However, the okhttp authors are probably better at tweaking this than me.